### PR TITLE
Milestone 3 : Game Logic

### PR DESCRIPTION
**Pull request by emmanuellekamwa**
Tic-Tac-Toe milestone 3: Game Logic
In this milestone, we:
-used classes to encapsulate the different parts of the game logic

### CODE REVIEW

File: lib/game.rb [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  end
+
+  # Display the game board
+  def display_board(board)
+    puts " #{board[0]}  | #{board[1]}  | #{board[2]}"
+    puts '-------------'
+    puts " #{board[3]}  | #{board[4]}  | #{board[5]}"
+    puts '-------------'
+    puts " #{board[6]}  | #{board[7]}  | #{board[8]}"
+  end
```
Comment (maelfosso) [inline lib/game.rb]: No `puts/gets` in any classes into `lib` folder

File: lib/game.rb [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+  # player-1 selects a token and the other token is assigned to player-2
+  def select_tokens(player_one, player_two, game_tokens)
+    puts "#{player_one.name}, Choose a token: 'X' or 'O'"
```
Comment (maelfosso) [inline lib/game.rb]: No puts/gets in any classes into lib folder

File: lib/game.rb [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  # returns board index where player's token is to be placed
+  def input_to_index(chosen_index, board)
+    # check token validity
+    valid = valid_move(chosen_index, board)
+
+    return chosen_index if valid
+  end
+
+  # checks if players made valid moves
+  def valid_move(index, _board)
```
Comment (maelfosso) [inline lib/game.rb]: why `_board` as parameter? It's never used in that method

File: lib/game.rb [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      unless game_tokens.include? item
+        full = false
+        break
+      end
+    end
+
+    full
+  end
+
+  def check_winning(player, board, winning_combination)
```
Comment (maelfosso) [inline lib/game.rb]: Please, remove `winning_combination` in parameters.
Move the `WINNING_COMBINATION` into that `Game` class as a constant

File: bin/main.rb [hunk 5]
```diff
@@ -1,76 +1,97 @@
...
+WINNING_COMBINATION = [
+  [0, 1, 2], # top row
+  [3, 4, 5], # middle row
+  [6, 7, 8], # bottom row
+  [0, 4, 8], # left to right diagonal
+  [2, 4, 6], # right to left diagonal
+  [0, 3, 6], # left-most column
+  [1, 4, 7], # middle column
+  [2, 5, 8] # right-most column
+].freeze
```
Comment (maelfosso) [inline bin/main.rb]: Move it into the `Game` class as a constant

File: bin/main.rb [hunk 6]
```diff
@@ -1,76 +1,97 @@
...
+
+board = %w[1 2 3 4 5 6 7 8 9]
```
Comment (maelfosso) [inline bin/main.rb]: move it into `Game` class as an attribute.
The Game cannot exist without a `board`. If we delete the board, the game doesn't have a reason to exist

File: bin/main.rb [hunk 7]
```diff
@@ -1,76 +1,97 @@
...
+game_tokens = %w[X O]
```
Comment (maelfosso) [inline bin/main.rb]: Move it also. Because it is here to only be sent to `Game` class method in parameter

File: lib/game.rb [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        player_one.token = selected_token
+        player_two.token = selected_token == game_tokens[0] ? game_tokens[1] : game_tokens[0]
+
+        puts 'Good choice!!'
+        puts "#{player_one.name}'s Token:  #{player_one.token}"
+        puts "#{player_two.name}'s Token:  #{player_two.token}"
+        break
+      end
+    end
+  end
```
Comment (maelfosso) [inline lib/game.rb]: Move it into the `bin/main.rb` file

File: lib/game.rb [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+  def check_board_full(board, game_tokens)
+    full = true
+
+    board.each do |item|
+      unless game_tokens.include? item
+        full = false
+        break
+      end
+    end
```
Comment (maelfosso) [inline lib/game.rb]: simply check is all the items in `board` are Numeric. That is
```rb
board.all?(Numeric)
```

File: lib/game.rb [hunk 10]
```diff
@@ -0,0 +1,102 @@
+#!/usr/bin/env ruby
+require_relative '../lib/player.rb'
+class Game
+  def create_player(name, token, move_count)
+    player = Player.new(name, token, move_count)
+    player
+  end
```
Comment (maelfosso) [inline lib/game.rb]: Why that line? 
Why not simply instantiate it in `bin/main.rb`?
There is apparently no reason for that method in that class

File: bin/main.rb [hunk 11]
```diff
@@ -1,76 +1,97 @@
...
+  # translate player's input into board index
+  board_index = game.input_to_index(position.to_i, board)
 
-  # if position is valid,
-  # update board with the current player's token
+  # update board with the current player's token at chosen board index
+  board_position = board_index.pred
+  board[board_position] = current_token
+  last_token = current_token
+  current_player.move_count += 1
```
Comment (maelfosso) [inline bin/main.rb]: Move it into the `Game` class wrapped into a method called for example `update_board`

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello,

Thank you for the job done so far
However, there are some missing requirement and/or unexpected behavior

1. There is a turn
```rb
alice's Turn (X)
 X  | 2  | 3
-------------
 4  | O  | 6
-------------
 7  | 8  | 9
Enter a position: 
2
 X  | X  | 3
-------------
 4  | O  | 6
-------------
 7  | 8  | 9
bob's Turn (O)
 X  | X  | 3
-------------
 4  | O  | 6
-------------
 7  | 8  | 9
Enter a position: 
3
 X  | X  | O
-------------
 4  | O  | 6
-------------
 7  | 8  | 9
alice's Turn (X)
 X  | X  | O
-------------
 4  | O  | 6
-------------
 7  | 8  | 9
Enter a position: 
4
 X  | X  | O
-------------
 X  | O  | 6
-------------
 7  | 8  | 9
There is a tie
```
The result is a tie when is not one. Kindly fix that
Same thing here
```rb
bob's Turn (O)
 1  | X  | X
-------------
 4  | O  | 6
-------------
 7  | 8  | 9
Enter a position: 
1
 O  | X  | X
-------------
 4  | O  | 6
-------------
 7  | 8  | 9
alice's Turn (X)
 O  | X  | X
-------------
 4  | O  | 6
-------------
 7  | 8  | 9
Enter a position: 
6
 O  | X  | X
-------------
 4  | O  | X
-------------
 7  | 8  | 9
There is a tie
```

2. There are some issues during the OOP implementation of your game. Kindly read the comments left below to know more about them

**Please, keep in mind that every code into the `lib` folder must be written in such a way that any other developer can use them to as a library to build a Tic-Tac-Toe game with a GUI or a Web application**
The code into the `lib` folder must be independent as much as possible from the user interface, any kind of user interface

Kindly make the suggested changes and submit a new code review request
Thanks

Mael FOSSO

File: lib/game.rb [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+
+  # returns board index where player's token is to be placed
+  def input_to_index(chosen_index, board)
+    # check token validity
+    valid = valid_move(chosen_index, board)
+
+    return chosen_index if valid
+  end
```
Comment (peterrobert) [inline lib/game.rb]: ### [OPTIONAL]

There should be no explicit return statement

**Review by peterrobert — CHANGES_REQUESTED** [review]
### CHANGES REQUIRED :recycle: :red_circle: 
### HELLO :wave: GOOD WORK SO FAR. GREAT WORK MAKING THE CHANGES AS REQUESTED BY THE PREVIOUS TSE :clap: HOWEVER, THERE ARE A FEW MORE CHANGES TO BE MADE KINDLY FIX THE FOLLOWING.

- [ ]  The application does not exist/stop due to invalid input. it informs the player about it and asks for their move again. If a player tries to move in an occupied position your application is stopping and it is throwing the following error :arrow_down:  Kindly fix this

`
update_board': undefined method pred  for nil:NilClass
`

### kindly fix this. If you have any question kindly reach me on slack @peterrobert

### HAPPY CODING :computer:

**Review by kitpao — CHANGES_REQUESTED** [review]
# 👋 Hello @emmanuellekamwa and @enkog 

👍 I loved that you added the option for the players to select a symbol to play with "X or O" 🌟 
I also see that you tried to fix the previous TSE suggestion, but you still have a bug there, please check it below:

- [x] It seems that any invalid position will be assigned to number 9 in your program, please fix this behavior: 
image

- [ ] **OPTIONAL** Currently, your program accepts 2 players with the same name, you could add a "2" to the name or ask for a new name to avoid this :)

### ♻️ Status: Your project requires some additional changes ♻️
➡️ Please request another review once you are ready 📝 
___
##### 👉 If you have any doubts you can contact me in slack as @kitpao
##### 🌟  If you have feedback about my reviews I would appreciate if can leave a recommendation in LinkedIn 💌 or let me know in Slack if you see improvements to be done 😛 

🔗 __[Github](https://github.com/kitpao) | [Twitter](https://twitter.com/Kitpao1) | [LinkedIn](https://www.linkedin.com/in/kitzia-paola-vidal/) | [Medium](https://medium.com/@kitpao91)__

**Review by peterrobert — CHANGES_REQUESTED** [review]
### CHANGES REQUIRED :recycle: :stop_sign: 
### HELLO :wave: GREAT PROGRESS SO FAR. YOUR PROJECT HAS IMPROVED BUT THERE ARE STILL SOME CHANGES TO BE DONE. KINDLY WORK ON THE FOLLOWING :arrow_down: 

- [x] When the game starts, it should show aboard in order for the player to know where exactly he should select. Kindly check on this :arrow_down: 

add board

- [ ] Great work the game does not stop when the same position is selected more than once. :+1:  However the game is not informing us that the position has already been selected. Kindly you need to add a message to inform us and asks the player to select another position:arrow_down:  

samePos

### if you have any question concerning the review kindly reach me on slack @peterrobert 

### HAPPY CODING :computer:

**Review by peterrobert — APPROVED** [review]
### APPROVED :+1: :100: 
### VERY NICE WORK MAKING THE CHANGES :+1: ITS A GO FOR ME :+1: 
### ON TO THE NEXT PROJECT AND HAPPY CODING

**Pull request merged**